KEP-4033: cgroup driver from CRI beta blog (#47054)
kubernetes/website: content/en/blog/_posts/2024-08-21-cgroup-driver-from-cri.md

---
layout: blog

> title: "Configuring Kubernetes just got easier: cgroup driver from CRI"

Reviewer: Also, this needs to follow title casing per [our style guide](https://kubernetes.io/docs/contribute/style/style-guide/)

Reviewer 2: ```suggestion
title: "Kubernetes v1.31: Autoconfiguration For Node Cgroup Driver (beta)"
```

> date: 2024-08-06

Reviewer: @a-mccarthy : Is this the correct date for publication? If not, please could you confirm what this needs to be?

Reviewer 2: Per the filename, I'm guessing it should be 2024-08-21

Reviewer 2: Per https://github.com/kubernetes/website/pull/47054#issuecomment-2253052256
```suggestion
date: 2024-08-21
```

Reviewer: Still hasn't been resolved in the commit visible to me. Please can we update it to the latest date?

Reviewer 2: /approve cancel

to highlight the problem

See https://github.com/kubernetes/website/pull/47054/files#r1720753283

> slug: cgroup driver from CRI to beta

Reviewer 2: ```suggestion
slug: cri-cgroup-driver-lookup-now-beta
```

author: >
  Peter Hunt (Red Hat)
---

> Historically, the cgroup driver has been a pain point for users running new

Reviewer 2: ```suggestion
Historically, configuring the correct cgroup driver has been a pain point for users running new
```

Kubernetes clusters. On Linux systems, there are two different cgroup drivers:
`cgroupfs` and `systemd`. In the past, both the [kubelet](/docs/reference/command-line-tools-reference/kubelet/)
and CRI implementation (like CRI-O or containerd) needed to be configured to use
the same cgroup driver, or else the kubelet would exit with an error. This was a
source of headaches for many cluster admins. However, there is light at the end of the tunnel!

> ## `KubeletCgroupDriverFromCRI` Feature Gate

Reviewer 2: ```suggestion
## Automated CRI driver configuration
```

or, less ideal
```suggestion
## `KubeletCgroupDriverFromCRI` feature gate
```

> In v1.28.0, the SIG Node community introduced the feature gate

Reviewer: ```suggestion
In v1.28.0, the SIG Node community introduced the [feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
```

`KubeletCgroupDriverFromCRI`, which instructs the kubelet to ask the CRI

> implementation which cgroup driver to use. It took a couple of releases for

Reviewer 2: ```suggestion
implementation which cgroup driver to use. A few minor releases of Kubernetes happened
whilst we waited for
```

support to land in the major two CRI implementations (containerd and CRI-O), but as of v1.31.0, this
feature is now beta!

In addition to setting the feature gate, a cluster admin needs to ensure their

> CRI implementatikn is new enough:

Reviewer: The current statement is likely to confuse users, as they might think they need to be on the latest version of the CRI for this to work. From the information given below, it would make more sense to say supported instead of new enough. 

```suggestion
CRI implementation is new enough:
```

> - containerd: Support has been added in v2.0.0
> - CRI-O: Support has been added in v1.28.0

Reviewer 2: ```suggestion
- containerd: Support was added in v2.0.0
- CRI-O: Support was added in v1.28.0
```

Then, they should ensure their CRI implementation is configured to the

> `cgroup_driver` they would like to use.

Reviewer 2: ```suggestion
cgroup driver they would like to use.
```

> ## Future Work

Reviewer 2: ```suggestion
## Future work
```

---
layout: blog
title: "Configuring Kubernetes just got easier: cgroup driver from CRI"

> date: 2024-07-01

Reviewer: Date potentially needs changing?

> slug: cgroup driver from CRI to betra

Reviewer: Typo
```suggestion
slug: cgroup driver from CRI to beta
```

author: >
  Peter Hunt (Red Hat)
---

> Historically, a pain point for users running new Kubernetes clusters has been

Reviewer: Converting to active voice, per [style guide](https://kubernetes.io/docs/contribute/style/style-guide/#use-active-voice). 

Historically, the cgroup driver has been a pain point for users running new Kubernetes clusters.

the cgroup driver. On Linux systems, there are two different cgroup drivers:

> `cgroupfs` and `systemd`. In the past, both the Kubelet and CRI implementation

Reviewer: ```suggestion
`cgroupfs` and `systemd`. In the past, both the [kubelet](/docs/reference/command-line-tools-reference/kubelet/) and CRI implementation
```

(like CRI-O or containerd) needed to be configured to use the same cgroup

> driver, or else the Kubelet would exit with an error. This was a source of

Reviewer: Do we want to include a description of the error here?
```suggestion
driver, or else the kubelet would exit with an error. This was a source of
```

headaches for many cluster admins. However, there is light at the end of the
tunnel!

## `KubeletCgroupDriverFromCRI` Feature Gate

> In v1.28.0, the SIG-Node community introduced the feature gate

Reviewer: [Style guide recommendation](https://kubernetes.io/docs/contribute/style/style-guide/#kubernetes-io-word-list)

```suggestion
In v1.28.0, the SIG Node community introduced the feature gate
```

> `KubeletCgroupDriverFromCRI`, which instructs the Kubelet to ask the CRI

Reviewer: ```suggestion
`KubeletCgroupDriverFromCRI`, which instructs the kubelet to ask the CRI
```

implementation which cgroup driver to use. It took a couple of releases for

> support to land in the major two CRI implementations, but as of v1.31.0, this

Reviewer: Specifying which CRI implementations would be beneficial here.
```suggestion
support to land in the major two CRI implementations, but as of v1.31.0, this
```

feature is now beta!

In addition to setting the feature gate, a cluster admin needs to ensure their
CRI implementation is new enough:

- containerd: Support has been added in v2.0.0
- CRI-O: Support has been added in v1.28.0

Then, they should ensure their CRI implementation is configured to the
`cgroup_driver` they would like to use.

## Future Work

> Eventually, support for the Kubelet's `cgroupDriver` configuration field will be

Reviewer: ```suggestion
Eventually, support for the kubelet's `cgroupDriver` configuration field will be
```

> dropped, and the Kubelet will fail to start if the CRI implementation isn't new

Reviewer: ```suggestion
dropped, and the kubelet will fail to start if the CRI implementation isn't new
```

> enough to have support for this feature. For now, the Kubelet's `cgroupDriver`

Reviewer: ```suggestion
enough to have support for this feature. For now, the kubelet's `cgroupDriver`
```

field is used as a fallback in case the CRI implementation is not new enough,

> and this behavior will also be dropped.

Reviewer: We're reiterating the same point made in the previous sentence. Any specific reason we're doing so?
